core/src/style/styleMixer.cpp

```diff
@@ -0,0 +1,222 @@
+#include "styleMixer.h"
+#include "style/style.h"
+#include "util/topologicalSort.h"
+#include "yaml-cpp/yaml.h"
+
+#include <algorithm>
+#include <set>
+
+namespace Tangram {
+
+std::vector<std::string> StyleMixer::getStylesToMix(const Node& _style) {
+
+    std::vector<std::string> names;
+
+    // 'base' style is the first item to mix.
+    if (const Node& base = _style["base"]) {
+        if (base.IsScalar()) { names.push_back(base.Scalar()); }
+    }
+
+    // 'mix' styles are mixed next, in order of declaration.
+    if (const Node& mix = _style["mix"]) {
+        if (mix.IsScalar()) {
+            names.push_back(mix.Scalar());
+        } else if (mix.IsSequence()) {
+            for (const auto& m : mix) {
+                if (m.IsScalar()) { names.push_back(m.Scalar()); }
+            }
+        }
+    }
+
+    return names;
+}
+
+std::vector<std::string> StyleMixer::getMixingOrder(const Node& _styles) {
+
+    // Input must be a map of names to style configuration nodes.
+    if (!_styles.IsMap()) {
+        return {};
+    }
+
+    // Dependencies are pairs of strings that form a DAG.
+    // If style 'a' mixes style 'b', the dependency would be {'b', 'a'}.
+    std::vector<std::pair<std::string, std::string>> dependencies;
+
+    for (const auto& entry : _styles) {
+        const auto& name = entry.first;
+        const auto& config = entry.second;
+        for (const auto& mix : getStylesToMix(config)) {
+            dependencies.push_back({ mix, name.Scalar() });
+        }
+    }
+
+    return topologicalSort(dependencies);
+}
+
+void StyleMixer::mixStyleNodes(Node _styles) {
+
+    // First determine the order of nodes to evaluate.
+    auto styleNamesSorted = getMixingOrder(_styles);
+
+    for (const auto& name : styleNamesSorted) {
+
+        const auto& style = _styles[name];
+
+        if (!style || !style.IsMap()) {
+            // Something's wrong here, try the next one!
+            continue;
+        }
+
+        // For each style to evaluate, get the list of styles that need to be mixed with this one.
+        auto stylesToMix = getStylesToMix(style);
+
+        std::vector<Node> mixins;
+        for (const auto& styleNameToMix : stylesToMix) {
+
+            // Skip mixing built-in styles.
+            const auto& builtIn = Style::builtInStyleNames();
+            if (std::find(builtIn.begin(), builtIn.end(), styleNameToMix) != builtIn.end()) {
+                continue;
+            }
+
+            mixins.push_back(_styles[styleNameToMix]);
+        }
+
+        applyStyleMixins(style, mixins);
```

Comment:
When mixing with base style, `mixins` is most of the time empty, we can save some allocs and only apply the `shaders:blocks:` mixing in that case.